### PR TITLE
Change NotInToString, IgnoredByEquals retention to compile time.

### PR DESCRIPTION
**Pull request by thespags**
Hi @alicederyn . I went to go try the new annotations but I have a unique situation in my project.

Scenario:
There is an interface Foo in module A.
With interface SubFoo that extends Foo in module B.

SubFoo is annotated with @FreeBuilder
Foo has some methods annotated with @NotInToString and/or @IgnoredByEquals.

Then when SubFoo is generated these annotations will not be retain across modules.

Let me know if the situation is strange, or if there is a reasons why one would want to limit the retention to source. Thank you.

### CODE REVIEW

**Comment by alicederyn** [comment]
Yikes, I forgot about inheritance. This change isn't going to be enough; you'll also need to:

- [ ] add the [Inherited](https://docs.oracle.com/javase/1.5.0/docs/api/java/lang/annotation/Inherited.html) annotation to NotInToString and IgnoredByEquals
- [ ] modify `Analyser` to check all superclasses and superinterfaces to see if any of them have the NotInToString/IgnoredByEquals
- [ ] add a bunch of tests

**Comment by alicederyn** [comment]
Thinking some more about this, this is why I didn't want new annotations in FreeBuilder :( The intent is for it to be a compile-time–only processor, never required by downstream libraries. These new annotations need to be moved to a separate library.

**Comment by alicederyn** [comment]
As an immediate fix, we need to change `Analyser` to ignore annotations on superclasses and superinterfaces.

**Comment by alicederyn** [comment]
Closing for now in favour of #416. We should discuss how to support inheritance in an issue.

**Comment by thespags** [comment]
A couple of responses and thoughts.

>  add the Inherited annotation to NotInToString and IgnoredByEquals

My understanding of `@Inherited` is that it only causes annotations to be inherited form super classes and it only works on classes. `NotInToString` and `IgnoredByEquals` are on methods only so it would have no affect.

> modify Analyser to check all superclasses and superinterfaces to see if any of them have the NotInToString/IgnoredByEquals

I'm not sure the concern here. Perhaps it's related to your subsequent comment "The current implementation behaves differently for incremental versus batch compilation." My local testing showed that those annotations work as expected (as long as the source and processing were together, I'll get back to that)

 > add a bunch of tests

Yup of course ::):.

To my original scenario. I think I'm trying to do something that isn't kosher. For more details I have a base interface, Node, that lives in Module A. I have another Module A' that generates sub interfaces of Node and puts them into Module B and then I use Freebuilder to build those sub Nodes in Module B. 

I have Module A' separated because it has dependencies I don't wish to expose to downstream users. And I have two layers of code generation, my own and Freebuilder which makes the mvn tasks a bit wonky so I have separate modules to make that cleaner.

In short, it sounds like my setup is not respecting how annotation processors should work.

**Comment by alicederyn** [comment]
i'll answer in pieces. You're right about Inherited. (What a bizarre annotation.)

**Comment by alicederyn** [comment]
During batch compilation, FreeBuilder will be handed many (possibly all) source files at once, meaning it can often see source-only annotations on all types in the current module. During incremental compilation (used by Eclipse and Gradle among others), FreeBuilder will be handed only a single source file, plus JAR-level information about supertypes and superinterfaces, meaning no access to source-only annotations on anything except the current compilation unit. In order to produce the same results in both modes, we need to pretend we can't see source-level annotations outside of the current compilation unit. This is the reason for #416.

**Comment by alicederyn** [comment]
>> modify Analyser to check all superclasses and superinterfaces to see if any of them have the NotInToString/IgnoredByEquals
>
> I'm not sure the concern here.

Basically, what happens if you declare the same method twice, and only annotate one of them `@IgnoredByEquals`? It's possible javax is resolving this for us...but more likely, we need to traverse the entire hierarchy and find all method declarations and check them all, lest we get unpredictable behaviour based on which method declaration javax hands us.

**Comment by alicederyn** [comment]
> I have two layers of code generation, my own and FreeBuilder

How is your code generation working? Is it an annotation processor?

> I have separate modules to make that cleaner

If you're writing a conventional annotation processor, you shouldn't need to do this: Java APT was designed to allow multiple annotation processors to interact within the same compilation run. Of course, it depends how the annotation processors are implemented! Let me know if you'd like to dig deeper into that.

**Comment by thespags** [comment]
>Basically, what happens if you declare the same method twice, and only annotate one of them @IgnoredByEquals?

I was making the assumption that if you did

```
public interface Foo {
 @IgnoredByEquals
 int property();
}
```
then
```
public interface FooExtends extends Foo {
 int property();
}
```
You should expect property to no longer be ignored. My quick interpretation is felt natural however, I could see that depending on how I or random developer was feeling you could have different understandings.

I tested that the following will start ignoring the property however I didn't do the inverse.
```
public interface Foo {
 int property();
}
```
then
```
public interface FooExtends extends Foo {
 @IgnoredByEquals
 int property();
}
```

TLDR: inheritance leads to confusion.

**Comment by thespags** [comment]
>How is your code generation working? Is it an annotation processor?

It's not an annotation processor. Basically the code generation is taking a protobuf representation where things are weakly referenced to a graph representation where things are strongly typed.

**Comment by alicederyn** [comment]
> I use Freebuilder to build those sub Nodes in Module B.

This is absolutely fine (if possibly unnecessary). The problem is just that the FreeBuilder JAR is a compile-only dependency. Suppose Bob is building their FooLibrary JAR, and using FreeBuilder to generate their builders; and suppose Clarissa is using FooLibrary in their BarServer. If FreeBuilder wasn't a compile-only dependency, Clarissa would be unintentionally running the FreeBuilder annotation processor when they build BarServer, silently disabling build system optimizations.

Now suppose `@NotInEquals` is changed to Class-level retention (the default, which this PR was doing), and FooLibrary starts using it. The annotation will now be referenced in the FooLibrary JAR, but without the FreeBuilder library, when BarServer loads FooLibrary, the JVM will complain that it doesn't know what NotInEquals was. Worse, tools (like Checkstyle) may refuse to run. If Bob leaves FreeBuilder as compile-only, Clarissa will need to start *explicitly* depending on FreeBuilder just to fix their toolchain!

If we want `@NotInEquals` to be inheritable, we need to move it into a separate library that's safe for  Bob to mark as a full dependency, that won't mess up Clarissa. Possibly a separate library that other code-generators could support 😁

**Comment by alicederyn** [comment]
(Apologies if all the Bob/Clarissa language comes across as patronizing by the way. I couldn't think of a way to explain myself clearly otherwise.)

**Comment by thespags** [comment]
@alicederyn No worries. It makes sense, I haven't had the need to keep an annotation retention source and your explanation makes it clear why it is needed with how things are set up today.

**Comment by alicederyn** [comment]
What if you have _two_ interfaces that declare foo(), one of them marked `@NotInEquals` and the other one not, and you inherit from both?

**Comment by alicederyn** [comment]
As a workaround, can you make your code generator redeclare the `@IgnoredByEquals`-annotated method on the interface it generates?

**Comment by thespags** [comment]
>What if you have two interfaces that declare foo(), one of them marked @NotInEquals and the other one not, and you inherit from both?

Yup great point.  That's why Java doesn't preserve annotations on class implementations which causes me to face palm every time I forget when I try to do some aspect on an interface. But I digress.

>As a workaround, can you make your code generator redeclare the @IgnoredByEquals-annotated method on the interface it generates?

Yup, I can do that. Thanks for all the feedback.

If it wasn't clear, I haven't been arguing for my PR, but ranting along that inheritance is the devil.

**Comment by alicederyn** [comment]
> If it wasn't clear, I haven't been arguing for my PR, but ranting along that inheritance is the devil.

It can be, definitely :)

**Comment by alicederyn** [comment]
And to be clear, I'd be happy to support a third-party effort to standardize annotations like this :)
